core/src/main/java/org/jruby/RubyObjectSpace.java

```diff
@@ -114,7 +114,7 @@ public static IRubyObject id2ref(IRubyObject recv, IRubyObject id) {
                 return object;
             } else {
                 runtime.getWarnings().warn("ObjectSpace is disabled; _id2ref only supports immediates, pass -X+O to enable");
-                return runtime.getNil();
+                throw recv.getRuntime().newRangeError(id+" This is not id value");
```

Comment:
Can you try to replicate the error message exactly? It should be `0x00000000000000 is not id value`, which you can probably make with `String.format('%016x is not id value', id)`.